test/unit/offset.js

```diff
@@ -568,4 +568,117 @@ QUnit.test( "iframe scrollTop/Left (see gh-1945)", function( assert ) {
 	}
 } );
 
+testIframe( "coordinates relative to document", "offset/rel-doc.html", function( assert, $, iframe, doc ) {
+	assert.expect( 768 );
+
+	var docElem = doc.documentElement, // <html> element
+		bodyElem = doc.body,
+		marker1 = doc.getElementById( "marker1" ),
+		marker2 = doc.getElementById( "marker2" ),
+		$marker1 = $( marker1 ),
+		$marker2 = $( marker2 ),
+
+		// Pixels
+		DOC_MARGIN = 1,
+		DOC_BORDER = 2,
+		DOC_PADDING = 4,
+		BODY_MARGIN = 8,
+		BODY_BORDER = 16,
+		BODY_PADDING = 32,
+		MARKER_TOP_LEFT = 64,
+		// Bit as Flag
+		FLAG_DOC_MARGIN = 1,
+		FLAG_DOC_BORDER = 2,
+		FLAG_DOC_PADDING = 4,
+		FLAG_BODY_MARGIN = 8,
+		FLAG_BODY_BORDER = 16,
+		FLAG_BODY_PADDING = 32;
+
+	// Test it in each situation that is made by combined properties.
+	function situations( affect, markerTopLeft, exBodyPadding, messageHead ) {
+		var flags, offset1, offset2, sumLen, labels, labelsInMessage;
+
+		function switchProp( elem, enable, styleProp, pixels, label ) {
+			if ( enable ) {
+				elem.style[ styleProp ] = pixels + "px";
+				if ( !exBodyPadding || elem !== bodyElem || styleProp !== "padding" ) {
+					sumLen += pixels;
+				}
+				labels.push( label );
+			} else {
+				elem.style[ styleProp ] = "0";
+			}
+		}
+
+		for ( flags = 0;
+				flags <= ( FLAG_DOC_MARGIN | FLAG_DOC_BORDER | FLAG_DOC_PADDING | FLAG_BODY_MARGIN | FLAG_BODY_BORDER | FLAG_BODY_PADDING );
+				flags++ ) {
```

Comment:
Ok, I try to rewrite tests.
You said that `situations()` should be called in each case of that 16 cases. Is my interpretation correct?

Also, that 16 cases include two cases in current code "CASE 1" and "CASE 2".
Is "CASE 3" in current code unnecessary?
Or, should two cases `marker1.style.position = "absolute"` and `marker1.style.position = "static"` be tested in each case of that 16 cases? That is, `situations()` is called in each case of 2x16 cases.